README
Router-R3 version 0.9.0
======================

    This mod is a XS wrapper around a C library R3.

    R3 is an URL router library with high performance, thus, it's implemented in C. It compiles your route paths into a prefix trie.

    By using the constructed prefix trie in the start-up time, you can dispatch routes with efficiency.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

SEE ALSO

The original C version L<"github repository"|https://github.com/c9s/r3> by L<c9s|https://metacpan.org/author/CORNELIUS>

This mod's L<"github repository"|https://github.com/CindyLinz/Perl-Router-R3>
All the source files with this mod are in the Router-R3 directory.


COPYRIGHT AND LICENCE

Put the correct copyright and licence information here.

Copyright (C) 2014 by Cindy Wang (CindyLinz)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8 or,
at your option, any later version of Perl 5 you may have available.




Changes
Revision history for Perl extension Router::R3.

0.11.1 2014.8.14
    To compatible with environment lack 'alloca()'

0.11.0 2014.8.12
    To compatible with perl 5.8, 5.10, 5.12.. (and also previously 5.14, 5.16, 5.18)

0.10.1 2014.6.3
    POD with benchmark (by c9s)

0.10.0 2014.6.3
    With r3 version 1.3.0
    Better error handling (by Perl die)

0.9.3 2014.5.23
    Change Makefile.PL for MacOS friendly (by c9s)

0.9.2 2014.5.22
    Fix version dependency perl 5.0182 -> 5.006
    Add -O2 to CCFLAGS

0.9.0 2014.5.21
    First release version ^_^


Makefile.PL
use 5.006;
use ExtUtils::MakeMaker;
use strict;
use warnings;

# See lib/ExtUtils/MakeMaker.pm for details of how to influence
# the contents of the Makefile that is written.
WriteMakefile(
    NAME              => 'Router::R3',
    VERSION_FROM      => 'lib/Router/R3.pm', # finds $VERSION
    PREREQ_PM         => {}, # e.g., Module::Name => 1.1
    ($] >= 5.005 ?     ## Add these new keywords supported since 5.005
      (ABSTRACT_FROM  => 'lib/Router/R3.pm', # retrieve abstract from module
       AUTHOR         => 'Cindy Wang (CindyLinz) <cindy@cpan.org>') : ()),
    CONFIGURE_REQUIRES => { 
        'ExtUtils::PkgConfig' => 0,
        'ExtUtils::MakeMaker' => 0,
    },
    CONFIGURE => sub {
        my @CCFLAGS = qw(-O2 -std=c99);
        my @LIBS = qw();
        if( exists &ExtUtils::PkgConfig::find ) {
            my %pkg_info = ExtUtils::PkgConfig->find('libpcre');
            die "libpcre is required." unless %pkg_info;
            print "Found libpcre:\n";
            print "modversion:  $pkg_info{modversion}\n";
            print "cflags:      $pkg_info{cflags}\n";
            print "libs:        $pkg_info{libs}\n";
            push @LIBS, $pkg_info{libs};
        } else {
            push @LIBS, '-lpcre';
        }

        my @INCLUDE_DIRS=qw(.);
        push @INCLUDE_DIRS, "/usr/local/include" if -e "/usr/local/include"; # check for homebrew path
        push @INCLUDE_DIRS, "/opt/local/include" if -e "/opt/local/include"; # check for macports path

        return {
            DEFINE            => '', # e.g., '-DHAVE_SOMETHING'
            LIBS    => \@LIBS,
            CCFLAGS => join(' ',@CCFLAGS),
            INC     => join(' ', map { "-I$_" } @INCLUDE_DIRS),
        };
    },
    # Un-comment this if you add C files to link with later:
    # OBJECT            => '$(O_FILES)', # link all the C files too
    # CCFLAGS           => join(' ', @CCFLAGS),
);
if  (eval {require ExtUtils::Constant; 1}) {
  # If you edit these definitions to change the constants used by this module,
  # you will need to use the generated const-c.inc and const-xs.inc
  # files to replace their "fallback" counterparts before distributing your
  # changes.
  my @names = (qw());
  ExtUtils::Constant::WriteConstants(
                                     NAME         => 'Router::R3',
                                     NAMES        => \@names,
                                     DEFAULT_TYPE => 'IV',
                                     C_FILE       => 'const-c.inc',
                                     XS_FILE      => 'const-xs.inc',
                                  );

}
else {
  use File::Copy;
  use File::Spec;
  foreach my $file ('const-c.inc', 'const-xs.inc') {
    my $fallback = File::Spec->catfile('fallback', $file);
    copy ($fallback, $file) or die "Can't copy $fallback to $file: $!";
  }
}


fill_r3_source.pl
#!/usr/bin/perl

use strict;
use warnings;

sub read_file {
    local $/;
    my($out, $f);
    open $f, $_[0];
    $out = <$f>;
    close $f;
    return $out;
}

sub write_file {
    my $f;
    open $f, ">$_[0]";
    print $f $_[1];
    close $f;
}

my $xs_code = read_file('R3.xs');

my $source_code = '';
$source_code =<<'END';
#define HAVE_STRNDUP
#define HAVE_STRDUP
END
for(qw(
    r3/3rdparty/zmalloc.h
    r3/include/r3_define.h
    r3/include/str_array.h
    r3/include/match_entry.h
    r3/include/r3.h
    r3/include/r3_list.h
    r3/include/r3_str.h
    r3/src/slug.h
    r3/3rdparty/zmalloc.c
    r3/src/match_entry.c
    r3/src/edge.c
    r3/src/list.c
    r3/src/node.c
    r3/src/str.c
    r3/src/token.c
    r3/src/slug.c
)) {
    print "fetching file $_\n";
    $source_code .= "/******* $_ *******/\n";
    $source_code .= read_file($_);
}

$source_code =~ s!(#include\s+".*)!/* $1 */!g;
# $source_code =~ s!\b(strn?dup)\b!my_$1!g;

$xs_code =~ s!(__R3_SOURCE_SLOT_BEGIN__.*?\n)(.*?)([^\n]*?__R3_SOURCE_SLOT_END__)!$1$source_code$3!s;

write_file('R3.xs', $xs_code);


scripts/routes.pl
#! /usr/bin/env perl
return {
"/foo/bar/baz" => 0,
"/foo/bar/qux" => 1,
"/foo/bar/quux" => 2,
"/foo/bar/corge" => 3,
"/foo/bar/grault" => 4,
"/foo/bar/garply" => 5,
"/foo/baz/bar" => 6,
"/foo/baz/qux" => 7,
"/foo/baz/quux" => 8,
"/foo/baz/corge" => 9,
"/foo/baz/grault" => 10,
"/foo/baz/garply" => 11,
"/foo/qux/bar" => 12,
"/foo/qux/baz" => 13,
"/foo/qux/quux" => 14,
"/foo/qux/corge" => 15,
"/foo/qux/grault" => 16,
"/foo/qux/garply" => 17,
"/foo/quux/bar" => 18,
"/foo/quux/baz" => 19,
"/foo/quux/qux" => 20,
"/foo/quux/corge" => 21,
"/foo/quux/grault" => 22,
"/foo/quux/garply" => 23,
"/foo/corge/bar" => 24,
"/foo/corge/baz" => 25,
"/foo/corge/qux" => 26,
"/foo/corge/quux" => 27,
"/foo/corge/grault" => 28,
"/foo/corge/garply" => 29,
"/foo/grault/bar" => 30,
"/foo/grault/baz" => 31,
"/foo/grault/qux" => 32,
"/foo/grault/quux" => 33,
"/foo/grault/corge" => 34,
"/foo/grault/garply" => 35,
"/foo/garply/bar" => 36,
"/foo/garply/baz" => 37,
"/foo/garply/qux" => 38,
"/foo/garply/quux" => 39,
"/foo/garply/corge" => 40,
"/foo/garply/grault" => 41,
"/bar/foo/baz" => 42,
"/bar/foo/qux" => 43,
"/bar/foo/quux" => 44,
"/bar/foo/corge" => 45,
"/bar/foo/grault" => 46,
"/bar/foo/garply" => 47,
"/bar/baz/foo" => 48,
"/bar/baz/qux" => 49,
"/bar/baz/quux" => 50,
"/bar/baz/corge" => 51,
"/bar/baz/grault" => 52,
"/bar/baz/garply" => 53,
"/bar/qux/foo" => 54,
"/bar/qux/baz" => 55,
"/bar/qux/quux" => 56,
"/bar/qux/corge" => 57,
"/bar/qux/grault" => 58,
"/bar/qux/garply" => 59,
"/bar/quux/foo" => 60,
"/bar/quux/baz" => 61,
"/bar/quux/qux" => 62,
"/bar/quux/corge" => 63,
"/bar/quux/grault" => 64,
"/bar/quux/garply" => 65,
"/bar/corge/foo" => 66,
"/bar/corge/baz" => 67,
"/bar/corge/qux" => 68,
"/bar/corge/quux" => 69,
"/bar/corge/grault" => 70,
"/bar/corge/garply" => 71,
"/bar/grault/foo" => 72,
"/bar/grault/baz" => 73,
"/bar/grault/qux" => 74,
"/bar/grault/quux" => 75,
"/bar/grault/corge" => 76,
"/bar/grault/garply" => 77,
"/bar/garply/foo" => 78,
"/bar/garply/baz" => 79,
"/bar/garply/qux" => 80,
"/bar/garply/quux" => 81,
"/bar/garply/corge" => 82,
"/bar/garply/grault" => 83,
"/baz/foo/bar" => 84,
"/baz/foo/qux" => 85,
"/baz/foo/quux" => 86,
"/baz/foo/corge" => 87,
"/baz/foo/grault" => 88,
"/baz/foo/garply" => 89,
"/baz/bar/foo" => 90,
"/baz/bar/qux" => 91,
"/baz/bar/quux" => 92,
"/baz/bar/corge" => 93,
"/baz/bar/grault" => 94,
"/baz/bar/garply" => 95,
"/baz/qux/foo" => 96,
"/baz/qux/bar" => 97,
"/baz/qux/quux" => 98,
"/baz/qux/corge" => 99,
"/baz/qux/grault" => 100,
"/baz/qux/garply" => 101,
"/baz/quux/foo" => 102,
"/baz/quux/bar" => 103,
"/baz/quux/qux" => 104,
"/baz/quux/corge" => 105,
"/baz/quux/grault" => 106,
"/baz/quux/garply" => 107,
"/baz/corge/foo" => 108,
"/baz/corge/bar" => 109,
"/baz/corge/qux" => 110,
"/baz/corge/quux" => 111,
"/baz/corge/grault" => 112,
"/baz/corge/garply" => 113,
"/baz/grault/foo" => 114,
"/baz/grault/bar" => 115,
"/baz/grault/qux" => 116,
"/baz/grault/quux" => 117,
"/baz/grault/corge" => 118,
"/baz/grault/garply" => 119,
"/baz/garply/foo" => 120,
"/baz/garply/bar" => 121,
"/baz/garply/qux" => 122,
"/baz/garply/quux" => 123,
"/baz/garply/corge" => 124,
"/baz/garply/grault" => 125,
"/qux/foo/bar" => 126,
"/qux/foo/baz" => 127,
"/qux/foo/quux" => 128,
"/qux/foo/corge" => 129,
"/qux/foo/grault" => 130,
"/qux/foo/garply" => 131,
"/qux/bar/foo" => 132,
"/qux/bar/baz" => 133,
"/qux/bar/quux" => 134,
"/qux/bar/corge" => 135,
"/qux/bar/grault" => 136,
"/qux/bar/garply" => 137,
"/qux/baz/foo" => 138,
"/qux/baz/bar" => 139,
"/qux/baz/quux" => 140,
"/qux/baz/corge" => 141,
"/qux/baz/grault" => 142,
"/qux/baz/garply" => 143,
"/qux/quux/foo" => 144,
"/qux/quux/bar" => 145,
"/qux/quux/baz" => 146,
"/qux/quux/corge" => 147,
"/qux/quux/grault" => 148,
"/qux/quux/garply" => 149,
"/qux/corge/foo" => 150,
"/qux/corge/bar" => 151,
"/qux/corge/baz" => 152,
"/qux/corge/quux" => 153,
"/qux/corge/grault" => 154,
"/qux/corge/garply" => 155,
"/qux/grault/foo" => 156,
"/qux/grault/bar" => 157,
"/qux/grault/baz" => 158,
"/qux/grault/quux" => 159,
"/qux/grault/corge" => 160,
"/qux/grault/garply" => 161,
"/qux/garply/foo" => 162,
"/qux/garply/bar" => 163,
"/qux/garply/baz" => 164,
"/qux/garply/quux" => 165,
"/qux/garply/corge" => 166,
"/qux/garply/grault" => 167,
"/quux/foo/bar" => 168,
"/quux/foo/baz" => 169,
"/quux/foo/qux" => 170,
"/quux/foo/corge" => 171,
"/quux/foo/grault" => 172,
"/quux/foo/garply" => 173,
"/quux/bar/foo" => 174,
"/quux/bar/baz" => 175,
"/quux/bar/qux" => 176,
"/quux/bar/corge" => 177,
"/quux/bar/grault" => 178,
"/quux/bar/garply" => 179,
"/quux/baz/foo" => 180,
"/quux/baz/bar" => 181,
"/quux/baz/qux" => 182,
"/quux/baz/corge" => 183,
"/quux/baz/grault" => 184,
"/quux/baz/garply" => 185,
"/quux/qux/foo" => 186,
"/quux/qux/bar" => 187,
"/quux/qux/baz" => 188,
"/quux/qux/corge" => 189,
"/quux/qux/grault" => 190,
"/quux/qux/garply" => 191,
"/quux/corge/foo" => 192,
"/quux/corge/bar" => 193,
"/quux/corge/baz" => 194,
"/quux/corge/qux" => 195,
"/quux/corge/grault" => 196,
"/quux/corge/garply" => 197,
"/quux/grault/foo" => 198,
"/quux/grault/bar" => 199,
"/quux/grault/baz" => 200,
"/quux/grault/qux" => 201,
"/quux/grault/corge" => 202,
"/quux/grault/garply" => 203,
"/quux/garply/foo" => 204,
"/quux/garply/bar" => 205,
"/quux/garply/baz" => 206,
"/quux/garply/qux" => 207,
"/quux/garply/corge" => 208,
"/quux/garply/grault" => 209,
"/corge/foo/bar" => 210,
"/corge/foo/baz" => 211,
"/corge/foo/qux" => 212,
"/corge/foo/quux" => 213,
"/corge/foo/grault" => 214,
"/corge/foo/garply" => 215,
"/corge/bar/foo" => 216,
"/corge/bar/baz" => 217,
"/corge/bar/qux" => 218,
"/corge/bar/quux" => 219,
"/corge/bar/grault" => 220,
"/corge/bar/garply" => 221,
"/corge/baz/foo" => 222,
"/corge/baz/bar" => 223,
"/corge/baz/qux" => 224,
"/corge/baz/quux" => 225,
"/corge/baz/grault" => 226,
"/corge/baz/garply" => 227,
"/corge/qux/foo" => 228,
"/corge/qux/bar" => 229,
"/corge/qux/baz" => 230,
"/corge/qux/quux" => 231,
"/corge/qux/grault" => 232,
"/corge/qux/garply" => 233,
"/corge/quux/foo" => 234,
"/corge/quux/bar" => 235,
"/corge/quux/baz" => 236,
"/corge/quux/qux" => 237,
"/corge/quux/grault" => 238,
"/corge/quux/garply" => 239,
"/corge/grault/foo" => 240,
"/corge/grault/bar" => 241,
"/corge/grault/baz" => 242,
"/corge/grault/qux" => 243,
"/corge/grault/quux" => 244,
"/corge/grault/garply" => 245,
"/corge/garply/foo" => 246,
"/corge/garply/bar" => 247,
"/corge/garply/baz" => 248,
"/corge/garply/qux" => 249,
"/corge/garply/quux" => 250,
"/corge/garply/grault" => 251,
"/grault/foo/bar" => 252,
"/grault/foo/baz" => 253,
"/grault/foo/qux" => 254,
"/grault/foo/quux" => 255,
"/grault/foo/corge" => 256,
"/grault/foo/garply" => 257,
"/grault/bar/foo" => 258,
"/grault/bar/baz" => 259,
"/grault/bar/qux" => 260,
"/grault/bar/quux" => 261,
"/grault/bar/corge" => 262,
"/grault/bar/garply" => 263,
"/grault/baz/foo" => 264,
"/grault/baz/bar" => 265,
"/grault/baz/qux" => 266,
"/grault/baz/quux" => 267,
"/grault/baz/corge" => 268,
"/grault/baz/garply" => 269,
"/grault/qux/foo" => 270,
"/grault/qux/bar" => 271,
"/grault/qux/baz" => 272,
"/grault/qux/quux" => 273,
"/grault/qux/corge" => 274,
"/grault/qux/garply" => 275,
"/grault/quux/foo" => 276,
"/grault/quux/bar" => 277,
"/grault/quux/baz" => 278,
"/grault/quux/qux" => 279,
"/grault/quux/corge" => 280,
"/grault/quux/garply" => 281,
"/grault/corge/foo" => 282,
"/grault/corge/bar" => 283,
"/grault/corge/baz" => 284,
"/grault/corge/qux" => 285,
"/grault/corge/quux" => 286,
"/grault/corge/garply" => 287,
"/grault/garply/foo" => 288,
"/grault/garply/bar" => 289,
"/grault/garply/baz" => 290,
"/grault/garply/qux" => 291,
"/grault/garply/quux" => 292,
"/grault/garply/corge" => 293,
"/garply/foo/bar" => 294,
"/garply/foo/baz" => 295,
"/garply/foo/qux" => 296,
"/garply/foo/quux" => 297,
"/garply/foo/corge" => 298,
"/garply/foo/grault" => 299,
"/garply/bar/foo" => 300,
"/garply/bar/baz" => 301,
"/garply/bar/qux" => 302,
"/garply/bar/quux" => 303,
"/garply/bar/corge" => 304,
"/garply/bar/grault" => 305,
"/garply/baz/foo" => 306,
"/garply/baz/bar" => 307,
"/garply/baz/qux" => 308,
"/garply/baz/quux" => 309,
"/garply/baz/corge" => 310,
"/garply/baz/grault" => 311,
"/garply/qux/foo" => 312,
"/garply/qux/bar" => 313,
"/garply/qux/baz" => 314,
"/garply/qux/quux" => 315,
"/garply/qux/corge" => 316,
"/garply/qux/grault" => 317,
"/garply/quux/foo" => 318,
"/garply/quux/bar" => 319,
"/garply/quux/baz" => 320,
"/garply/quux/qux" => 321,
"/garply/quux/corge" => 322,
"/garply/quux/grault" => 323,
"/garply/corge/foo" => 324,
"/garply/corge/bar" => 325,
"/garply/corge/baz" => 326,
"/garply/corge/qux" => 327,
"/garply/corge/quux" => 328,
"/garply/corge/grault" => 329,
"/garply/grault/foo" => 330,
"/garply/grault/bar" => 331,
"/garply/grault/baz" => 332,
"/garply/grault/qux" => 333,
"/garply/grault/quux" => 334,
"/garply/grault/corge" => 335,

"/" => 336,

"/post/{year}/{month}" => 337,
};


scripts/benchmark.pl
#!/usr/bin/env perl
use lib 'lib';
use strict;
use warnings;
use Time::HiRes qw(time);
use Router::R3;
use Router::Boom;
use Router::Boom::Method;
use Router::Simple;
use HTTP::Router;
use Benchmark qw(:all);

$|++;


my @Benchmarks = (
    { desc => "plain string matching", path => '/corge/quux/bar' },
    { desc => "regexp string matching", path => '/post/2012/03' },
    { desc => "first charactar matching", path => '/' },
);

my $routes = require "scripts/routes.pl";

my $router_r3 = Router::R3->new($routes);

my $router_simple = Router::Simple->new();
for my $path ( keys %$routes )  {
    $router_simple->connect($path, { val => $routes->{ $path } } );
}


my $router_boom = Router::Boom->new();
for my $path ( keys %$routes )  {
    $router_boom->add($path, $routes->{ $path } );
}

my $http_router = HTTP::Router->new;
for my $path ( keys %$routes )  {
    $http_router->add_route($path => ( ));
}


my $N = 500000;



foreach my $b (@Benchmarks) {
    printf "Benchmarking '%s' by path '%s'\n", $b->{desc}, $b->{path};
    printf "===============================================================\n";

    my $path = $b->{path};

=pod
    my ($s, $d);

    printf "% 20s - ", "Router::R3";
    $s = time();
    for ( 1..$N ) {
        $router_r3->match($path);
    }
    $d = time() - $s;
    printf "%10.2f i/s\n", ($N / $d);

    printf "% 20s - ", "Router::Simple";
    $s = time();
    for ( 1..$N ) {
        $router_simple->match($path);
    }
    $d = time() - $s;
    printf "%10.2f i/s\n", ($N / $d);


    printf "% 20s - ", "HTTP::Router";
    $s = time();
    for ( 1..$N ) {
        $http_router->match($path);
    }
    $d = time() - $s;
    printf "%10.2f i/s\n", ($N / $d);


    printf "% 20s - ", "Router::Boom";
    $s = time();
    for ( 1..$N ) {
        $router_boom->match($path);
    }
    $d = time() - $s;
    printf "%10.2f i/s\n", ($N / $d);
=cut

    cmpthese(
        -1, {
            'Router::Simple'         => sub { $router_simple->match($path) },
            'Router::Boom'           => sub { $router_boom->match($path) },
            'Router::R3'             => sub { $router_r3->match($path) },
            'HTTP::Router'           => sub { $http_router->match($path) },
        }
    );
    printf "\n";
}


1;
__END__

Benchmark Result

https://gist.github.com/c9s/488d6fd4e53c34758559

Benchmarking 'plain string matching' by path '/corge/quux/bar'
===============================================================
                   Rate  HTTP::Router Router::Simple  Router::Boom    Router::R3
HTTP::Router      505/s            --           -88%         -100%         -100%
Router::Simple   4225/s          738%             --          -98%         -100%
Router::Boom   177535/s        35090%          4102%            --          -80%
Router::R3     877196/s       173773%         20660%          394%            --
 
Benchmarking 'regexp string matching' by path '/post/2012/03'
===============================================================
                   Rate  HTTP::Router Router::Simple  Router::Boom    Router::R3
HTTP::Router      383/s            --           -88%         -100%         -100%
Router::Simple   3199/s          734%             --          -97%          -99%
Router::Boom   107789/s        28007%          3269%            --          -67%
Router::R3     329098/s        85715%         10187%          205%            --
 
Benchmarking 'first charactar matching' by path '/'
===============================================================
                    Rate  HTTP::Router Router::Simple Router::Boom    Router::R3
HTTP::Router       179/s            --           -91%        -100%         -100%
Router::Simple    2036/s         1035%             --         -99%         -100%
Router::Boom    153597/s        85476%          7443%           --          -88%
Router::R3     1310719/s       730158%         64266%         753%            --


lib/Router/R3.pm
package Router::R3;

use 5.006;
use strict;
use warnings;
use Carp;

require Exporter;
use AutoLoader;

our $VERSION = '0.011001';

sub AUTOLOAD {
    # This AUTOLOAD is used to 'autoload' constants from the constant()
    # XS function.

    my $constname;
    our $AUTOLOAD;
    ($constname = $AUTOLOAD) =~ s/.*:://;
    croak "&Router::R3::constant not defined" if $constname eq 'constant';
    my ($error, $val) = constant($constname);
    if ($error) { croak $error; }
    goto &$AUTOLOAD;
}

require XSLoader;
XSLoader::load('Router::R3', $VERSION);

# Preloaded methods go here.

# Autoload methods go after =cut, and are processed by the autosplit program.

1;
__END__
# Below is stub documentation for your module. You'd better edit it!

=head1 NAME

Router::R3 - URL router library with high performance

=head1 SYNOPSIS

  use Router::R3;

  my $router = Router::R3->new(
    '/static/index.html' => 1,
    '/post/{id}' => 2,
    '/post_comment/{id:\d+}/{id2}' => 3,
  );

  my($match, $captures);
  ($match, $captures) = $router->match('/static/index.html'); # (1, {})
  ($match, $captures) = $router->match('/post/123'); # (2, { id => '123' })
  ($match, $captures) = $router->match('/post_comment/123/456'); # (3, { id => '123', id2 => '456' })
  ($match, $captures) = $router->match('/post_comment/xxx/456'); # ()  no match

  # or you can pass a hashref or an arrayref when Router::R3->new

  my $router = Router::R3->new(['/static', 1, '/post/{id}', 2]);
  my $router = Router::R3->new({'/static' => 1, '/post/{id}' => 2});

  # The latter one of each rule could be any perl scalar
  #  It'll be given to you when the rule is matched.
  #  It's better not to put anything which is treated as false here.

=head1 DESCRIPTION

This mod is a XS wrapper around a C library R3.

R3 is an URL router library with high performance, thus, it's implemented in C. It compiles your route paths into a prefix trie.

By using the constructed prefix trie in the start-up time, you can dispatch routes with efficiency.

=head2 PATTERN SYNTAX

  /blog/post/{id}        use [^/]+ regular expression by default.
  /blog/post/{id:\d+}    use `\d+` regular expression instead of default.
  /blog/post/{id:\d{2}}  use `\d{2}` regular expression instead of default.

=head1 INSTALLATION

Install from CPAN:

    $ cpanm Router::R3

Or from GitHub:

    $ git clone https://github.com/CindyLinz/Perl-Router-R3.git
    $ cd Perl-Router-R3
    $ cpanm -v --installdeps .

=head2 METHODS

=over 4

=item $router = Router::R3->new(...)

    The constructor

=item ($matched, \%captures) = $router->match($test_string)

    Match strings

=back

=head1 Benchmark

    Benchmarking 'plain string matching' by path '/corge/quux/bar'
    ===============================================================
                    Rate  HTTP::Router Router::Simple  Router::Boom    Router::R3
    HTTP::Router      203/s            --           -89%         -100%         -100%
    Router::Simple   1782/s          779%             --          -99%         -100%
    Router::Boom   168658/s        83094%          9365%            --          -82%
    Router::R3     954407/s       470684%         53461%          466%            --

    Benchmarking 'regexp string matching' by path '/post/2012/03'
    ===============================================================
                    Rate  HTTP::Router Router::Simple  Router::Boom    Router::R3
    HTTP::Router     1076/s            --           -88%          -99%         -100%
    Router::Simple   9309/s          765%             --          -91%          -97%
    Router::Boom   104387/s         9602%          1021%            --          -66%
    Router::R3     306925/s        28426%          3197%          194%            --

    Benchmarking 'first charactar matching' by path '/'
    ===============================================================
                        Rate  HTTP::Router Router::Simple Router::Boom    Router::R3
    HTTP::Router      3839/s            --           -87%         -98%         -100%
    Router::Simple   30545/s          696%             --         -83%          -98%
    Router::Boom    180555/s         4603%           491%           --          -88%
    Router::R3     1535999/s        39910%          4929%         751%            --


=head1 SEE ALSO

The original C version L<"github repository"|https://github.com/c9s/r3> by L<c9s|https://metacpan.org/author/CORNELIUS>

This mod's L<"github repository"|https://github.com/CindyLinz/Perl-Router-R3>
All the source files with this mod are in the Router-R3 directory.

=head1 AUTHOR

Cindy Wang (CindyLinz)

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2014 by Cindy Wang (CindyLinz)

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8 or,
at your option, any later version of Perl 5 you may have available.


=cut
